easy_logging.go

```diff
@@ -0,0 +1,138 @@
+	i.everTriedToInitialize = false
+	i.initParameter = ""
+	i.configureCounter = 0
+}
+
+func initEasyLogging(configFilePathFromConnectionString string) error {
+	if !allowedToInitialize(configFilePathFromConnectionString) {
+		return nil
+	}
+	config, err := getClientConfig(configFilePathFromConnectionString)
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		easyLoggingInitTrials.setInitTrial(configFilePathFromConnectionString)
+		return nil
+	}
+	var logLevel string
+	logLevel, err = getLogLevel(config.Common.LogLevel)
+	if err != nil {
+		return err
+	}
+	var logPath string
+	logPath, err = getLogPath(config.Common.LogPath)
+	if err != nil {
+		return err
+	}
+	reconfigureEasyLogging(logLevel, logPath)
+	easyLoggingInitTrials.setInitTrial(configFilePathFromConnectionString)
+	easyLoggingInitTrials.increaseReconfigureCounter()
+	return nil
+}
+
+func reconfigureEasyLogging(logLevel string, logPath string) error {
+	err := logger.SetLogLevel(logLevel)
+	if err != nil {
+		return err
+	}
+	var output io.Writer
+	output, err = createLogWriter(logPath)
+	if err != nil {
+		return err
+	}
+	logger.SetOutput(output)
+	return nil
+}
+
+func createLogWriter(logPath string) (io.Writer, error) {
+	if strings.EqualFold(logPath, "STDOUT") {
+		return os.Stdout, nil
+	}
+	logFileName := path.Join(logPath, "snowflake.log")
+	file, err := os.OpenFile(logFileName, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
+	if err != nil {
+		return nil, err
+	}
+	return io.MultiWriter(file, os.Stdout), nil
+}
+
+func allowedToInitialize(configFilePathFromConnectionString string) bool {
+	triedToInitializeWithoutConfigFile := easyLoggingInitTrials.everTriedToInitialize && easyLoggingInitTrials.initParameter == ""
+	isAllowedToInitialize := !easyLoggingInitTrials.everTriedToInitialize || (triedToInitializeWithoutConfigFile && configFilePathFromConnectionString != "")
+	if !isAllowedToInitialize && easyLoggingInitTrials.initParameter != configFilePathFromConnectionString {
+		logger.Warnf("Easy logging will not be configured for CLIENT_CONFIG_FILE=%s because it was previously configured for a different client config", configFilePathFromConnectionString)
+	}
+	return isAllowedToInitialize
+}
+
+func getLogLevel(logLevel string) (string, error) {
+	if logLevel == "" {
+		logger.Warn("LogLevel in client config not found. Using default value: OFF")
+		return levelOff, nil
+	}
+	return toLogLevel(logLevel)
+}
+
+func getLogPath(logPath string) (string, error) {
+	logPathOrDefault := logPath
+	if logPath == "" {
+		logger.Warn("LogPath in client config not found. Using temporary directory as a default value")
+		logPathOrDefault = os.TempDir()
+	}
+	pathWithGoSubdir := path.Join(logPathOrDefault, "go")
+	exists, err := existsDir(pathWithGoSubdir)
+	if err != nil {
+		return "", err
+	}
+	if !exists {
+		err = os.MkdirAll(pathWithGoSubdir, 0755)
+		if err != nil {
+			return "", err
+		}
+	}
+	return pathWithGoSubdir, nil
+}
+
+func existsDir(dirPath string) (bool, error) {
```

Comment:
Same function exists in tests, maybe we can remove it?

Reply:
in what file? I don't see it.

Reply:
In `client_configuration.go` :P The implementation for dir and file is the same, right?

Reply:
they are different. In dirExists we additionally check if it is a dictionary.
If it is a file, but not a dictionary we return false.
If we used existsFile function we would get true.